src/query/plan/rule_based_planner.hpp

```diff
@@ -569,62 +583,55 @@ class RuleBasedPlanner {
-      }
+    std::vector<Symbol> new_symbols;
+    std::unordered_set<Symbol> expand_symbols(bound_symbols.begin(), bound_symbols.end());
 
-      // This should never happen
-      MG_ASSERT(false, "Pattern filter does not exist!");
-    }
+    auto filters = matching.filters;
 
-    return nullptr;
-  }
+    std::unordered_map<Symbol, std::vector<Symbol>> named_paths;
 
-  std::unique_ptr<LogicalOperator> MakeExistsFilter(const Exists &exists, AstStorage &storage,
-                                                    const SymbolTable &symbol_table,
-                                                    const std::unordered_set<Symbol> &bound_symbols) {
-    auto *from_node = static_cast<NodeAtom *>(exists.pattern_->atoms_[0]);
-    auto *relationship = static_cast<EdgeAtom *>(exists.pattern_->atoms_[1]);
-    auto *to_node = static_cast<NodeAtom *>(exists.pattern_->atoms_[2]);
-    auto node1_symbol = symbol_table.at(*from_node->identifier_);
-    auto node2_symbol = symbol_table.at(*to_node->identifier_);
-    auto edge_symbol = symbol_table.at(*relationship->identifier_);
-    auto direction = relationship->direction_;
+    last_op = GenExpand(std::move(last_op), matching.expansions[0], symbol_table, expand_symbols, matching, storage,
+                        filters, named_paths, new_symbols, storage::View::OLD);
 
-    std::vector<storage::EdgeTypeId> edge_types;
-    edge_types.reserve(relationship->edge_types_.size());
-    for (const auto &type : relationship->edge_types_) {
-      edge_types.push_back(GetEdgeType(type));
-    }
+    last_op = std::make_unique<Limit>(std::move(last_op), storage.Create<PrimitiveLiteral>(1));
 
-    std::vector<Symbol> once_symbols(bound_symbols.begin(), bound_symbols.end());
-    std::unique_ptr<LogicalOperator> last_op = std::make_unique<Once>(once_symbols);
+    last_op = std::make_unique<EvaluatePatternFilter>(std::move(last_op), matching.symbol.value());
 
-    last_op = std::make_unique<Expand>(std::move(last_op), node1_symbol, node2_symbol, edge_symbol, direction,
-                                       edge_types, false, storage::View::OLD);
+    return last_op;
+  }
 
-    last_op = std::make_unique<Limit>(std::move(last_op), storage.Create<PrimitiveLiteral>(1));
+  std::vector<std::shared_ptr<LogicalOperator>> ExtractPatternFilters(Filters &filters, const SymbolTable &symbol_table,
+                                                                      AstStorage &storage,
+                                                                      const std::unordered_set<Symbol> &bound_symbols) {
+    std::vector<std::shared_ptr<LogicalOperator>> operators;
 
-    last_op = std::make_unique<EvaluatePatternFilter>(std::move(last_op), symbol_table.at(exists));
+    for (const auto &filter : filters) {
+      for (const auto &matching : filter.matchings) {
+        switch (matching.type) {
+          case PatternFilterType::EXISTS: {
```

Comment:
I would add here default so compailer doesn't scream in case we add new value here. I think it should prevent it.

```default:
        break;
```

Reply:
I would avoid adding default, because my reasoning for opt-in to switch is exactly to make the compiler scream at us if we don't handle an enum value somewhere. I think for long time now the `default` branches are prohibited in our codebase (we didn't updated it retrospectively, but we are trying to catch them in PRs).

Reply:
Basically without default the compiler enforces us to think how to handle each enum value in places where the enum is used.

Reply:
That makes sense, I didn't think about that from such a point of view. Thanks for clarifying :smile: